### PR TITLE
Nutanix AHV Fence agent

### PR DESCRIPTION
**Pull request by nxgovind**
Fence agent support for Nutanix AHV clusters managed by Prism Central
    Fence agent requires Nutanix API v4 support, this means that
    Prism central version should support API v4.

### CODE REVIEW

File: agents/ahv/fence_ahv.py [hunk 1]
```diff
@@ -0,0 +1,688 @@
+"""
+AHV Fence agent
+
+Compatible with AHV v3 and v4 API
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: Please remove this comment for now, until V3 is addressed!

File: agents/ahv/fence_ahv.py [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+import sys
+import time
+import uuid
+import requests
+
+sys.path.append("@FENCEAGENTSLIBDIR@")
+import fencing
+
+
+V4_VERSION = '4.0.b1'
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: This will need to be updated to the release version!

File: agents/ahv/fence_ahv.py [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+                Args:
+                url(str): URL of the API end point.
+                method(str): HTTP request method.
+                headers(dict): HTTP request headers
+                **kwargs: Any other keyword args relevant to the method.
+
+                Returns:
+                response: response object.
+                """
+                requests.packages.urllib3.disable_warnings()
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: please refer to my comment in the previous patch about this!

File: agents/ahv/fence_ahv.py [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+                        session.headers.update(headers)
+
+                response = None
+
+                try:
+                        response = session.request(method, url, **kwargs)
+                        response.raise_for_status()
+                except requests.exceptions.RequestException as err:
+                        logging.error("API call failed: %s", response.text)
+                        logging.error("Error message: %s", err)
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: you can combine these two into one error message

File: agents/ahv/fence_ahv.py [hunk 5]
```diff
@@ -0,0 +1,688 @@
...
+                        raise NutanixClientException(f"API call failed: {err}") from err
+                if response.status_code not in self.valid_status_codes:
+                        logging.error("API call returned status code %s", response.status_code)
+                        raise NutanixClientException(f"API call failed: {response}")
+
+                return response
+
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: remove extra line

File: agents/ahv/fence_ahv.py [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+
+class V4NutanixClient(NutanixClient):
+        """
+        V4 Client
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: maybe a bit more descriptive docstring here!

File: agents/ahv/fence_ahv.py [hunk 7]
```diff
@@ -0,0 +1,688 @@
...
+                Returns:
+                dict: HTTP headers.
+                """
+                resp = None
+                headers = {'Accept':'application/json',
+                           'Content-Type': 'application/json'}
+
+                if vm_uuid:
+                        try:
+                                resp = self._get_vm(vm_uuid)
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: fix the indentation!

File: agents/ahv/fence_ahv.py [hunk 8]
```diff
@@ -0,0 +1,688 @@
...
+                        except AHVFenceAgentException as err:
+                                logging.error("Unable to retrieve etag")
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: ditto

File: agents/ahv/fence_ahv.py [hunk 9]
```diff
@@ -0,0 +1,688 @@
...
+                Get a list all registered VMs.
+
+                Args:
+                filter_str(str): filter string.
+
+                Returns:
+                str: A json formatted output of details of all VMs.
+                """
+                vm_url = self.vm_url
+                # Fix this
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: address this plz or if it's already addressed, remove this comment

File: agents/ahv/fence_ahv.py [hunk 10]
```diff
@@ -0,0 +1,688 @@
...
+                        headers['Ntnx-Request-Id'] = request_id
+
+                return headers
+
+        def _get_all_vms(self, filter_str=None, limit=None):
+                """
+                Get a list all registered VMs.
+
+                Args:
+                filter_str(str): filter string.
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: add docstring for limit

and also a more descriptive docstring please

File: agents/ahv/fence_ahv.py [hunk 11]
```diff
@@ -0,0 +1,688 @@
...
+                elif filter_str and not limit:
+                        vm_url = f"{vm_url}?$filter={filter_str}"
+                elif limit and not filter_str:
+                        vm_url = f"{vm_url}?$limit={limit}"
+
+                logging.info("Sending GET request to get VM details, %s", vm_url)
+                header_str = self._get_headers()
+
+                try:
+                        resp = self.request(url=vm_url, method='GET',
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: fix indent! default should be 4
this should apply to the entire file! unless the upstream uses 8 by default

File: agents/ahv/fence_ahv.py [hunk 12]
```diff
@@ -0,0 +1,688 @@
...
+
+                vm_url = self.vm_url + f"/{vm_uuid}"
+                logging.info("Sending GET request to get VM details, %s", vm_url)
+                header_str = self._get_headers()
+
+                try:
+                        resp = self.request(url=vm_url, method='GET',
+                                            headers=header_str, verify=self.verify)
+                except NutanixClientException as err:
+                        logging.error("Failed to retrieve VM details"
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: nit: leave the space at the end of this line, so next line can start without the space

e.g: logging.error("Failed to retrieve VM details "
                                "for VM UUID: vm_uuid")

File: agents/ahv/fence_ahv.py [hunk 13]
```diff
@@ -0,0 +1,688 @@
...
+                Returns:
+                object: HTTP response object.
+                """
+                if not vm_uuid:
+                        logging.error("VM UUID was not provided")
+                        raise AHVFenceAgentException("VM UUID not provided")
+
+                vm_url = self.vm_url + f"/{vm_uuid}"
+                logging.info("Sending GET request to get VM details, %s", vm_url)
+                header_str = self._get_headers()
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: get_headers call self._get_vm on line 110 so how does this work?
am I missing something?

File: agents/ahv/fence_ahv.py [hunk 14]
```diff
@@ -0,0 +1,688 @@
...
+
+                if not vm_uuid:
+                        logging.error("VM UUID was not provided")
+                        raise AHVFenceAgentException("VM UUID not provided")
+
+                vm_url = self.vm_url + f"/{vm_uuid}/$actions/power-off"
+                logging.info("Sending request to power off VM, %s", vm_url)
+
+                try:
+                        headers_str = self._get_headers(vm_uuid)
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: I understand why we need headers and we need to do a VM get! but here seems like for any power op API, we need to do a VM get call. I was thinking to cache the header but if there is a VM update then the etag will be different. let's think to see if there is any possible ways to reduce the number of API calls!

File: agents/ahv/fence_ahv.py [hunk 15]
```diff
@@ -0,0 +1,688 @@
...
+
+                task_url = f"{self.task_url}/{task_uuid}"
+                header_str = self._get_headers()
+                task_resp = None
+                interval = 10
+
+                if not timeout:
+                        timeout = MIN_TIMEOUT
+
+                try:
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: you can move this under the while statement, there no need of doing this additional API call. you can just call the API and if the status is succeeded then break, if not keep trying

File: agents/ahv/fence_ahv.py [hunk 16]
```diff
@@ -0,0 +1,688 @@
...
+                Returns:
+                str: A json formatted response to the HTTP request.
+                """
+                resp = None
+
+                if not vm_name and not vm_uuid:
+                        logging.error("Require at least one of VM name or VM UUID")
+                        raise AHVFenceAgentException("No arguments provided")
+
+                if not vm_uuid:
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: when does this happen? 
I asked this in the previous patch. I'm just curious to know how fence agent queries for the VM. is it based on VM name or their UUId? I assume it's their name?

File: agents/ahv/fence_ahv.py [hunk 17]
```diff
@@ -0,0 +1,688 @@
...
+                power_state = resp.json()['data']['powerState']
+
+                if power_state.lower() != 'on':
+                        power_state = 'off'
+
+                return power_state
+
+        def set_power_state(self, vm_name=None, vm_uuid=None,
+                            power_state='off', timeout=None):
+                """
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: add args and description to the dostring

Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: ditto for all the functions in this file

File: agents/ahv/fence_ahv.py [hunk 18]
```diff
@@ -0,0 +1,688 @@
...
+        try:
+                client.list_vms(limit=1)
+        except AHVFenceAgentException as err:
+                logging.error("Connection to Prism Central Failed")
+                logging.error(err)
+                fencing.fail(fencing.EC_LOGIN_DENIED)
+
+        return client
+
+def list_vms(client, options):
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: This name is confusing, if fence gent requires to have this name, then please change the name of the function under the class!

File: agents/ahv/fence_ahv.py [hunk 19]
```diff
@@ -0,0 +1,688 @@
...
+            options(dict): options dictionary
+
+        Returns:
+            None
+        """
+        vmid = None
+        name = None
+        power_state = None
+
+        if "--uuid" in options:
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: I'm a little confused here. Like I asked above, does fence agent uses UUID or VM-name to perform ops?

File: agents/ahv/fence_ahv.py [hunk 20]
```diff
@@ -0,0 +1,688 @@
...
+                logging.error("Need VM name or VM UUID to set power cycling a VM")
+                fencing.fail(fencing.EC_GENERIC_ERROR)
+
+        try:
+                client.power_cycle_vm(vm_name=name, vm_uuid=vmid, timeout=timeout)
+        except AHVFenceAgentException as err:
+                logging.error(err)
+                fencing.fail(fencing.EC_GENERIC_ERROR)
+
+def define_new_opts():
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py:551]: what is the purpose of this function? is it required?

File: agents/ahv/fence_ahv.py [hunk 21]
```diff
@@ -0,0 +1,688 @@
...
+        fencing.all_opt["power_timeout"]["default"] = str(MIN_TIMEOUT)
+        options = fencing.check_input(device_opt, fencing.process_input(device_opt))
+        docs = {}
+        docs["shortdesc"] = "Fencing agent for Nutanix AHV Cluster VMs."
+        docs["longdesc"] = """
+                            fence_ahv is a fencing agent for nodes
+                            deployed on Nutanix AHV cluster with AHV cluster
+                            being managed by Prism Central.
+                            """
+        docs["vendorurl"] = "https://www.nutanix.com"
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: should the URL be the doc ref for the V4 API? or is it just a vendor URL?

**Review by amireibagi — CHANGES_REQUESTED** [review]

File: agents/ahv/fence_ahv.py (hunk 2 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Done. I think I may have to test with 4.0.b1/b2 if the current master does not have 4.0 version string.

File: agents/ahv/fence_ahv.py (hunk 3 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Yes, makes sense. Removing this now. Any warnings thrown at the customer end might not affect functionality but just the user experience.

File: agents/ahv/fence_ahv.py (hunk 4 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: I would like these to be in two different lines. If I have to combine these two then I may have to add a new line character in the string.

File: agents/ahv/fence_ahv.py (hunk 5 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Since the return statement is separate from the try/except block I generally leave a line as per PEP8 recommendations.

File: agents/ahv/fence_ahv.py (hunk 6 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Thanks, I forgot to elaborate some docstrings. I'll revisit all of them.

File: agents/ahv/fence_ahv.py (hunk 7 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: The current indentation standard with fence agents are 8 spaces. So I have aligned with that requirement.

File: agents/ahv/fence_ahv.py (hunk 9 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Yes, this is already fixed. I have removed the comment.

File: agents/ahv/fence_ahv.py (hunk 8 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: I ran pylint with 8 spaces and it returns no errors.

File: agents/ahv/fence_ahv.py (hunk 11 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Yes, upstream is using 8 spaces.

File: agents/ahv/fence_ahv.py (hunk 12 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Done

File: agents/ahv/fence_ahv.py (hunk 15 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Good catch, not sure how I missed the break statement. Added break if task has succeeded.

File: agents/ahv/fence_ahv.py (hunk 16 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Fence agent caller will execute the fence_ahv.py and pass the required parameters. But the VM identifier can be either a VM name or VM uuid based on what identifier was used by the customer to configure the fence agent. So either one can be passed by pacemaker.

File: agents/ahv/fence_ahv.py (hunk 18 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Done. Changed the outer function to get_list()

File: agents/ahv/fence_ahv.py (hunk 20 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py:551]: This is for fencing.py. We need to define some options that can be set by pacemaker during configuration.

File: agents/ahv/fence_ahv.py (hunk 21 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Just the vendor..

File: agents/ahv/fence_ahv.py (hunk 4 above)
Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: Ack

File: agents/ahv/fence_ahv.py (hunk 7 above)
Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: Ack

File: agents/ahv/fence_ahv.py (hunk 8 above)
Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: Ack

File: agents/ahv/fence_ahv.py (hunk 11 above)
Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: Ack

File: agents/ahv/fence_ahv.py (hunk 13 above)
Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: I see that you are calling self._get_headers() without vm_uuid, make sense!

File: agents/ahv/fence_ahv.py (hunk 14 above)
Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: just FYI! something to think about!

File: agents/ahv/fence_ahv.py (hunk 15 above)
Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: Still you are calling this api twice here! you can simply start your loop, call the api and check for the task status, if it succeeded then break if not keep retrying! you don't need line 337!

File: agents/ahv/fence_ahv.py (hunk 16 above)
Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: I see, thanks

**Review by amireibagi — CHANGES_REQUESTED** [review]

File: agents/ahv/fence_ahv.py (hunk 14 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Yes, we can try to optimize this. I started thinking around this but any solution I could think of became too complex. But we can optimize it further.

File: agents/ahv/fence_ahv.py (hunk 15 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: I see what you are saying. Earlier, I tried to optimize it by letting the code not go through the while loop if possible. I have made the change. I'll update the PR.

File: agents/ahv/fence_ahv.py [hunk 22]
```diff
@@ -0,0 +1,744 @@
...
+                Args:
+                url(str): URL of the API end point.
+                method(str): HTTP request method.
+                headers(dict): HTTP request headers
+                **kwargs: Any other keyword args relevant to the method.
+
+                Returns:
+                response: response object.
+                """
+                session = requests.Session()
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: do we want to create a session on every request?

File: agents/ahv/fence_ahv.py [hunk 23]
```diff
@@ -0,0 +1,744 @@
...
+                session.auth = (self.username, self.password)
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: same here! we can move both these line into init, if new session is not required for every request

File: agents/ahv/fence_ahv.py (hunk 5 above)
Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: Ack

File: agents/ahv/fence_ahv.py [hunk 24]
```diff
@@ -0,0 +1,744 @@
...
+                                logging.error("Unable to retrieve task status")
+                                raise AHVFenceAgentException from err
+
+                        if task_status == 'SUCCEEDED':
+                                break
+
+                        time.sleep(interval)
+                        timeout = timeout - interval
+
+                        if task_status == 'SUCCEEDED':
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: Why do we have this line duplicated? this is the same as line 349!

Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Thanks, I missed it. I think I might have mistakenly copied twice.

File: agents/ahv/fence_ahv.py (hunk 22 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: If I move the session into __init__() then I keep getting rate limit error.

requests.exceptions.HTTPError: 429 Client Error: TOO MANY REQUESTS for url: https://10.101.63.173:9440/api/vmm/v4.0/ahv/config/vms?$filter=name%20eq%20'TestVM1'

Am I missing something? Is there a way to avoid this error and preserve the session?

Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: Ah I see, this is probably referring to the rate limit. This means that server is limiting the number of requests you can make over a specified time period. I guess we can add retry to that logic but for now this should be ok as long as the stress testing won't cause any issue

File: agents/ahv/fence_ahv.py [hunk 25]
```diff
@@ -0,0 +1,744 @@
...
+        Exception class for AHV Fence Agent
+        """
+
+
+class NutanixClient:
+        """
+        Nutanix REST Client
+        """
+        def __init__(self, username, password):
+                self.username = username
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: Docstring for this

File: agents/ahv/fence_ahv.py [hunk 26]
```diff
@@ -0,0 +1,755 @@
...
+    """
+    def __init__(self, username, password):
+        """
+        Init method
+
+        Args:
+            username(str): username for Prism Central user
+            password(str): password for Prism Central user account
+
+        Returns:
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: Nit: we don't return anything in the Init method!

File: agents/ahv/fence_ahv.py [hunk 27]
```diff
@@ -0,0 +1,755 @@
...
+            logging.error("API call failed: %s", response.text)
+            logging.error("Unknown error %s", err)
+            raise NutanixClientException(f"API call failed: {err}") from err
+
+        if response.status_code not in self.valid_status_codes:
+            logging.error("API call returned status code %s", response.status_code)
+            raise NutanixClientException(f"API call failed: {response}")
+
+        return response
+
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py:77]: nit: remove extra line

File: agents/ahv/fence_ahv.py [hunk 28]
```diff
@@ -0,0 +1,755 @@
...
+        vm_url = self.vm_url
+
+        if filter_str and limit:
+            vm_url = f"{vm_url}?$filter={filter_str}&$limit={limit}"
+        elif filter_str and not limit:
+            vm_url = f"{vm_url}?$filter={filter_str}"
+        elif limit and not filter_str:
+            vm_url = f"{vm_url}?$limit={limit}"
+
+        logging.info("Sending GET request to get VM details, %s", vm_url)
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: Nit: update the message which refers to the list of VMs. This debug message is the same as line 229 so when it comes down to debugging it may cause some confusion!

File: agents/ahv/fence_ahv.py [hunk 29]
```diff
@@ -0,0 +1,755 @@
...
+            object: HTTP response object.
+        """
+        resp = None
+
+        if not vm_uuid:
+            logging.error("VM UUID was not provided")
+            raise AHVFenceAgentException("VM UUID not provided")
+
+        vm_url = self.vm_url + f"/{vm_uuid}/$actions/power-off"
+        logging.info("Sending request to power off VM, %s", vm_url)
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: Nit: I would recommend to use the vm_uuid here. perhaps in the request method you can log the actual URL which is being sent to the PC. but you can leav ethat one as log.debug not info. This way we won't flood the logs with every API call!

Comment (amireibagi) [reply agents/ahv/fence_ahv.py]: ditto for the rest of the debug messages

File: agents/ahv/fence_ahv.py [hunk 30]
```diff
@@ -0,0 +1,755 @@
...
+
+        try:
+            resp = self._get_vm(vm_uuid)
+        except AHVFenceAgentException as err:
+            logging.error("Unable to retrieve power state of VM %s", vm_uuid)
+            raise AHVFenceAgentException from err
+
+        power_state = resp.json()['data']['powerState']
+
+        if power_state.lower() != 'on':
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: we could have a pause state! not common but if acropolis is down I believe the VM can be in paused state!

File: agents/ahv/fence_ahv.py (hunk 28 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py:175]: I have 2 spaces between classes as per PEP8 style guide

File: agents/ahv/fence_ahv.py (hunk 27 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py:77]: I have 2 spaces between classes as per PEP8 style guide

File: agents/ahv/fence_ahv.py (hunk 30 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Yes, I agree. In this case I am only interested if the VM is powered-on or not.

File: agents/ahv/fence_ahv.py (hunk 29 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Done

File: agents/ahv/fence_ahv.py (hunk 28 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: I have changed all other occurrences to vm_uuid but this one. Here, there is no vm_uuid being passed since, this method just returns all VMs by applying any filter string that was passed.

File: agents/ahv/fence_ahv.py [hunk 31]
```diff
@@ -0,0 +1,561 @@
...
+
+        return vm_uuid
+
+    def _get_vm(self, vm_uuid):
+        if not vm_uuid:
+            logging.error("VM UUID was not provided")
+            raise AHVFenceAgentException("VM UUID not provided")
+
+        vm_url = self.vm_url + f"/{vm_uuid}"
+        logging.info("Sending GET request to get VM details, %s", vm_uuid)
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: this is the same message as line 113, maybe rewording this could help during debugging

File: agents/ahv/fence_ahv.py [hunk 32]
```diff
@@ -0,0 +1,561 @@
...
+
+        if not timeout:
+            timeout = MIN_TIMEOUT
+        else:
+            try:
+                timeout = int(timeout)
+            except ValueError as err:
+                timeout = MIN_TIMEOUT
+
+        while task_status != 'SUCCEEDED':
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py:272]: maybe you should redesign the condition here. 
we are already checking task_status == 'SUCCEEDED' on line 256, after this line the task status is not changing right?
so there is no point in having while task_status != 'SUCCEEDED'
you can either have this as while True and have couple break statement like:
1. if task is succeeded 
2. if timeout has reached 

or you can have the loop condition as if timeout has reached!

File: agents/ahv/fence_ahv.py [hunk 33]
```diff
@@ -0,0 +1,561 @@
...
+                logging.error("Unable to retrieve UUID of VM, %s", vm_name)
+                raise AHVFenceAgentException from err
+
+        try:
+            resp = self._get_vm(vm_uuid)
+        except AHVFenceAgentException as err:
+            logging.error("Unable to retrieve power state of VM %s", vm_uuid)
+            raise AHVFenceAgentException from err
+
+        power_state = resp.json()['data']['powerState']
```
Comment (amireibagi) [inline agents/ahv/fence_ahv.py]: why not moving these under the try catch statement?
Just want to make sure in case of a empty response (not sure it can happen) but atleast we are protected and correct error is being reported, this applies to other method in this file!

**Review by amireibagi — CHANGES_REQUESTED** [review]

File: agents/ahv/fence_ahv.py (hunk 33 above)
Comment (nxgovind) [reply agents/ahv/fence_ahv.py]: Done

**Review by amireibagi — APPROVED** [review]

**Review by amireibagi — APPROVED** [review]

**Review by amireibagi — APPROVED** [review]
